core/field_number.js

```diff
@@ -135,6 +135,9 @@ Blockly.FieldNumber.prototype.getNumRestrictor = function(opt_min, opt_max,
   if (this.negativeAllowed_) {
     pattern += "|[-]";
   }
+  if (this.exponentialAllowed_) {
+    pattern += "|e";
+  }
   return new RegExp(pattern);
```

Comment:
This looks good! One thing to note is that Scratch 2.0/3.0 handle 'e' in a case-insensitive way when evaluating inputs. This is because the ECMAScript `Number()` function is case-insensitive when dealing with 'e'.

It seems a little inconvenient to need to remember whether caps-lock is on when inputting a number... especially since Scratch tries hard to hide case from the user.

Since it's not just 'e', but all of Scratch, that's case sensitive, I think it's best to change the RegEx flag rather than just a component of the regex, to make changes involving letters faster in future.

![I tested both capital and lowercase 'e's using the join block and they are equivelant](https://user-images.githubusercontent.com/18113170/48307276-5a92b880-e541-11e8-9c8f-70b69f039e73.png)

Warning: the suggestion below is untested
```suggestion
  return new RegExp(pattern, 'i');
```

Reply:
Yes, I thought of this. While testing in 2.0 I noticed that if you try to type `3E4`, it will automatically replace the `E` as you press it with `e`. That's a bit wonky to me; I think just allowing capital E makes more sense. I'll leave it up to a PR reviewer to decide, though (but yeah, your change should work; otherwise, appending `|[eE]` to the regex will work too).

Reply:
Just a note that it's probably not too hard to match 2.0's behaviour if that's what's decided, according to [Blockly documentation](https://developers.google.com/blockly/reference/js/Blockly.FieldNumber) (my emphasis):

> An optional function that is called to validate any constraints on what the user entered. Takes the new text as an argument and returns either the accepted text, ***a replacement text***, or null to abort the change.

The advantage of 2.0's way is that a person who is not familiar with scientific notation will not be confused by the (lack of a) distinction in "5E3" and "5e3".